src/function/cast_string_non_nested_functions.cpp

```diff
@@ -83,5 +88,151 @@ bool TryCastStringToTimestamp::tryCast<timestamp_sec_t>(const char* input, uint6
+            auto fieldType = inferMinimalTypeFromString(splitEle[1]);
+            fields.emplace_back(fieldKey, std::move(fieldType));
+        }
+        return LogicalType::STRUCT(std::move(fields));
+    } else {
+        return LogicalType::STRING();
+    }
+}
+
+LogicalType inferMinimalTypeFromString(const std::string& str) {
+    constexpr char array_begin = common::CopyConstants::DEFAULT_CSV_LIST_BEGIN_CHAR;
+    constexpr char array_end = common::CopyConstants::DEFAULT_CSV_LIST_END_CHAR;
+    auto cpy = StringUtils::ltrim(StringUtils::rtrim(str));
+    StringUtils::toUpper(cpy);
+    if (cpy.size() == 0) {
+        return LogicalType::ANY();
+    }
+    // Boolean
+    if (cpy == "TRUE" || cpy == "FALSE") {
+        return LogicalType::BOOL();
+    }
+    // the reason we're not going to try to match to a minimal width integer
+    // is because if we're infering the type of integer from a sequence of
+    // increasing integers, we're bound to underestimate the width
+    // if we only sniff the first few elements
+    // integer
+    if (RE2::FullMatch(cpy, "(-?0)|(-?[1-9]\\d*)")) {
+        if (cpy.size() >= 1 + NumericLimits<int128_t>::digits()) {
+            return LogicalType::DOUBLE();
+        }
+        int128_t val;
+        if (!trySimpleInt128Cast(cpy.c_str(), cpy.length(), val)) {
+            return LogicalType::STRING();
+        }
+        if (NumericLimits<int64_t>::isInBounds(val)) {
+            return LogicalType::INT64();
+        }
+        return LogicalType::INT128();
+    }
+    // Real value checking
+    if (RE2::FullMatch(cpy, "(\\+|-)?(0|[1-9]\\d*)?\\.(\\d*)")) {
+        if (cpy[0] == '-') {
+            cpy.erase(cpy.begin());
+        }
+        if (cpy.size() <= DECIMAL_PRECISION_LIMIT) {
+            auto decimalPoint = cpy.find('.');
+            KU_ASSERT(decimalPoint != std::string::npos);
+            return LogicalType::DECIMAL(cpy.size() - 1, cpy.size() - decimalPoint - 1);
+        } else {
+            return LogicalType::DOUBLE();
+        }
+    }
+    // date
+    if (isDate(cpy)) {
+        return LogicalType::DATE();
+    }
+    // it might just be quicker to try cast to timestamp
+    timestamp_t tmp;
+    if (common::Timestamp::tryConvertTimestamp(cpy.c_str(), cpy.length(), tmp)) {
+        return LogicalType::TIMESTAMP();
+    }
+
+    // UUID
+    if (isUUID(cpy)) {
+        return LogicalType::UUID();
+    }
+
+    // interval checking
+    if (isInterval(cpy)) {
+        return LogicalType::INTERVAL();
+    }
+
+    if (cpy.front() == array_begin && cpy.back() == array_end) {
+        auto split = StringUtils::split(cpy.substr(1, cpy.size() - 2), ",", false);
+        auto childType = LogicalType::ANY();
+        for (auto& ele : split) {
+            childType = LogicalTypeUtils::combineTypes(childType, inferMinimalTypeFromString(ele));
+        }
+        return LogicalType::LIST(std::move(childType));
+    }
+
+    if (cpy.front() == '{' && cpy.back() == '}') {
```

Comment:
is this correct? What if the infermapOrStruct fails?
If i have a string like: "{abc}", will this be interpreted as string or struct?

Reply:
It'll give a string.